File: java/src/main/java/org/sammancoaching/Phonebook.java
package org.sammancoaching;

import java.util.*;

public class Phonebook {
    private final Map<String, String> storage;

    public Phonebook() {
        this.storage = new HashMap<>();
    }

    public void add(String name, String number) {
        storage.put(name, number);
    }

    public String lookup(String name) {
        if (!storage.containsKey(name)) {
            throw new RuntimeException("unknown name: " + name);
        }
        return storage.get(name);
    }

    public int size() {
        return storage.size();
    }


    public List<Map.Entry<String, String>> findClashes(String numberToCheck) {
        List<Map.Entry<String, String>> entries = new ArrayList<>();

        for (Map.Entry<String, String> entry : storage.entrySet()) {
            String number = entry.getValue();
            if (number.startsWith(numberToCheck) || numberToCheck.startsWith(number)) {
                entries.add(entry);
            }
        }

        return entries;
    }
}

File: kotlin/src/main/kotlin/org/sammancoaching/phonebook/Phonebook.kt
package org.sammancoaching.phonebook

class Phonebook {
    private val storage: MutableMap<String?, String?> = HashMap<String?, String?>()

    fun add(name: String?, number: String?) {
        storage[name] = number
    }

    fun lookup(name: String): String {
        if (!storage.containsKey(name)) {
            throw RuntimeException("unknown name: " + name)
        }
        return storage[name]!!
    }

    fun size(): Int {
        return storage.size
    }

    fun findClashes(numberToCheck: String): MutableList<MutableMap.MutableEntry<String?, String?>> {
        val entries: MutableList<MutableMap.MutableEntry<String?, String?>> = ArrayList()

        for (entry in storage.entries) {
            val number: String = entry.value!!
            if (number.startsWith(numberToCheck) || numberToCheck.startsWith(number)) {
                entries.add(entry)
            }
        }

        return entries
    }
}